File: 배열/leetcode1472(Medium)-Design Browser History.py
# 202ms(99.30%), 16.8MB(10.57%)
class BrowserHistory:

    def __init__(self, homepage: str):
        self.visited_urls = [homepage]
        self.current_index = 0

    def visit(self, url: str) -> None:
        self.visited_urls = self.visited_urls[0 : self.current_index + 1]
        self.visited_urls.append(url)
        self.current_index += 1

    def back(self, steps: int) -> str:
        self.current_index = max(self.current_index - steps, 0)
        return self.visited_urls[self.current_index]

    def forward(self, steps: int) -> str:
        self.current_index = min(self.current_index + steps, len(self.visited_urls) - 1)
        return self.visited_urls[self.current_index]

File: 배열/leetcode1472(Medium)-Design Browser History.ts
// 178ms(81.11%), 52MB(82.22%)
class BrowserHistory {
  visitedUrls: string[]
  currentIndex: number

  constructor(homepage: string) {
    this.visitedUrls = [homepage]
    this.currentIndex = 0
  }

  visit(url: string): void {
    this.visitedUrls = this.visitedUrls.slice(0, this.currentIndex + 1)
    this.visitedUrls.push(url)
    this.currentIndex += 1
  }

  back(steps: number): string {
    this.currentIndex = Math.max(this.currentIndex - steps, 0)
    return this.visitedUrls[this.currentIndex]
  }

  forward(steps: number): string {
    this.currentIndex = Math.min(
      this.currentIndex + steps,
      this.visitedUrls.length - 1,
    )
    return this.visitedUrls[this.currentIndex]
  }
}
